src/main/java/macrobase/analysis/stats/MinCovDet.java

```diff
@@ -148,19 +149,34 @@ private RealVector getMean(List<? extends HasMetrics> data) {
     private List<MetricsWithScore> findKClosest(int k, List<? extends HasMetrics> data) {
         List<MetricsWithScore> scores = new ArrayList<>();
 
+        double[] scoreVec = new double[data.size()];
+
         for (int i = 0; i < data.size(); ++i) {
             HasMetrics d = data.get(i);
-            scores.add(new MetricsWithScore(d.getMetrics(),
-                                            getMahalanobis(mean, inverseCov, d.getMetrics())));
+            Double score = getMahalanobis(mean, inverseCov, d.getMetrics());
+            scores.add(new MetricsWithScore(d.getMetrics(), score));
+            scoreVec[i] = score;
         }
 
         if (scores.size() < k) {
             return scores;
         }
 
-        scores.sort((a, b) -> a.getScore().compareTo(b.getScore()));
+        double cutoff = QuickSelect.select(scoreVec, k, scoreVec.length);
+
+        List<MetricsWithScore> ret = new ArrayList<>(k);
 
-        return scores.subList(0, k);
+        for(MetricsWithScore mws : scores) {
```

Comment:
I think you're right, but QuickSelect.select() operates on the array of doubles, not the `scores` array. I think the best way would be to write a QuickSelect implementation that operates on Java comparables so we could do this and also avoid a copy. I wasn't going to do it this time around, but maybe it's worth it.